### PR TITLE
feat: add function for obtaining a list of holders in a given block number

### LINKED ISSUE #296: [FEATURE] (New Feature) Function for obtaining a list of holders in a given block number
## Is your feature request related to a problem? Please describe.
We want to obtain a list of token holders in a particular block number in the past.

E.g. pay dividends to holders in a specific block number.

### New API: Extraction of the list of holders in a specific block number (POST)
Batch processing execution instructions

- IN: list_id (unique ID to be assigned to each list), block_number

### New API: retrieving the list of holders in a specific block number (GET)
- IN: list_id
- OUT: processing status, list data

### PR DESCRIPTION
**Pull request by purplesmoke05**
Close #296

### POST

```
/token/holders/{token_address}/collection
```
* Added POST endpoint.
* This api expect “contract_address” in path, “list_id” and “block_number” in body.
* It also requires "issuer-address" in header.
* ”list_id” is assumed to be UUID.
* If existing combinations of block_number and contract_address is given, this api will return existing list_id.

### GET

```
/token/holders/{token_address}/collection/{list_id}
```
* This api expect “contract_address” and “list_id “ in path.
* It will return dict having keys "status" and "holders".

## 2. Batch feature

* It syncs all events from 0 to given block number.
* If it has already collected events from 0 to some block number and stored data in DB, it can reuse the data.
* Token holders address is excluding issuer address as we cannot get totalSupply at specific block_number(If all events we may fetch, it could).

## 3. others

* I wonder that I should add required field("issuer-address") in header or not.

### CODE REVIEW

File: app/routers/token_holders.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+router = APIRouter(
+    prefix="/token",
+    tags=["token"],
+)
+
+# POST: /token/holders/{token_address}/collection
+
+
+@router.post(
```
Comment (urushibata-k) [inline app/routers/token_holders.py:56]: needless newline

File: app/routers/token_holders.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        .filter(Token.issuer_address == issuer_address)
+        .filter(Token.token_status != 2)
+    )
+    _token = query.first()
+    if _token is None:
+        raise HTTPException(status_code=404, detail="token not found")
+    if _token.token_status == 0:
+        raise InvalidParameterError("wait for a while as the token is being processed")
+
+    # Validate list id
```
Comment (urushibata-k) [inline app/routers/token_holders.py:89]: Format check should be done in model class.

File: batch/indexer_token_holders.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+        :param block_from: Block from
+        :param block_to: Block to
+        :return: None
+        """
+        try:
+            # Get "HolderChanged" events from exchange contract
+            exchange_contract = ContractUtils.get_contract(contract_name="IbetExchangeInterface", contract_address=self.tradable_exchange_address)
+
+            holder_changed_events = exchange_contract.events.HolderChanged.getLogs(fromBlock=block_from, toBlock=block_to)
```
Comment (urushibata-k) [inline batch/indexer_token_holders.py:226]: use ContractUtils.get_event_logs

File: app/model/db/token_holders.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+"""
+from enum import Enum
+
+from sqlalchemy import Column, String, BigInteger
+
+from .base import Base
+
+
+class TokenHoldersList(Base):
+    """Issued Token"""
```
Comment (YoshihitoAso) [inline app/model/db/token_holders.py:27]: コメント誤りですね

File: app/model/db/token_holders.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class TokenHolderBatchStatus(str, Enum):
+    PENDING = "pending"
+    DONE = "done"
+    FAILED = "failed"
+
+
+class TokenHolder(Base):
+    """Issued Token"""
```
Comment (YoshihitoAso) [inline app/model/db/token_holders.py:49]: こちらもコメントの誤りです

File: app/routers/token_holders.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    "/holders/{token_address}/collection",
+    response_model=CreateTokenHoldersListResponse,
+    responses=get_routers_responses(422, 404, InvalidParameterError),
+)
+def create_collection(
+    data: CreateTokenHoldersListRequest,
+    token_address: str = Path(
+        ...,
+        example="0xABCdeF1234567890abcdEf123456789000000000",
+        description="UUID v4 required",
```
Comment (YoshihitoAso) [inline app/routers/token_holders.py:66]: おそらく説明が誤りですかね

File: app/routers/token_holders.py (hunk 1 above)
Comment (purplesmoke05) [reply app/routers/token_holders.py:55]: Thanks, removed needless lines.

File: app/routers/token_holders.py (hunk 2 above)
Comment (purplesmoke05) [reply app/routers/token_holders.py:87]: I see. I made validation in model class.

File: batch/indexer_token_holders.py (hunk 3 above)
Comment (purplesmoke05) [reply batch/indexer_token_holders.py]: A value exchange_contract is stored when “__load_token_info” is called, so I fixed to reuse it.
And I also removed code in tests which registers token to token_list contract.

File: app/model/db/token_holders.py (hunk 4 above)
Comment (purplesmoke05) [reply app/model/db/token_holders.py]: 失礼いたしました。修正しました。

File: app/model/db/token_holders.py (hunk 5 above)
Comment (purplesmoke05) [reply app/model/db/token_holders.py]: こちらも修正いたしました。

File: app/routers/token_holders.py (hunk 6 above)
Comment (purplesmoke05) [reply app/routers/token_holders.py]: path名（token_address）で何者か判別できるため、descriptionを削除しました。

File: batch/indexer_token_holders.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+            tmp_events = []
+            try:
+                # Get "HolderChanged" events from exchange contract
+                holder_changed_events = ContractUtils.get_event_logs(
+                    contract=self.exchange_contract,
+                    event="HolderChanged",
+                    block_from=block_from,
+                    block_to=block_to,
+                )
+            except ABIEventFunctionNotFound:
```
Comment (YoshihitoAso) [inline batch/indexer_token_holders.py:220]: get_event_logsはABIEventFunctionNotFoundが発生した場合は、空のリストを返すので、この制御は不要ですね。

**Review by YoshihitoAso — APPROVED** [review]

**Pull request merged**